src/ge/mod.rs

```diff
@@ -1,59 +1,451 @@
+    // why is `.double()` internal only?
+    let r1 = Scalar::from(2u8) * &p1;
+
+    *r = Ge25519::from(&r1);
 }
 
-pub unsafe extern "C" fn ge25519_mul8() {
+/// r = [8]P
+//ge25519_double_partial(r, t);
+//ge25519_double_partial(r, r);
+//ge25519_double(r, r);
+#[no_mangle]
+pub unsafe extern "C" fn ge25519_mul8(r: *mut Ge25519, p: *const Ge25519) {
+    let p1 = match EdwardsPoint::try_from(&*p) {
+        Ok(v) => v,
+        Err(_) => return,
+    };
+
+    let r1 = p1.mul_by_cofactor();
 
+    *r = Ge25519::from(&r1);
 }
 
-pub unsafe extern "C" fn ge25519_double_scalarmult_vartime() {
+/// r = [s1]p1 + [s2]base
+#[no_mangle]
+pub unsafe extern "C" fn ge25519_double_scalarmult_vartime(
+    r: *mut Ge25519,
+    p1: *const Ge25519,
+    s1: *const Bignum256Modm,
+    s2: *const Bignum256Modm,
+) {
+    let p1 = match EdwardsPoint::try_from(&*p1) {
+        Ok(v) => v,
+        Err(_) => return,
+    };
 
+    let (s1, s2) = (
+        Scalar::from_unpacked_u32(*s1),
+        Scalar::from_unpacked_u32(*s2),
+    );
+
+    let r1 = s1 * p1 + &ED25519_BASEPOINT_TABLE * &s2;
+
+    *r = Ge25519::from(&r1);
 }
 
-pub unsafe extern "C" fn ge25519_double_scalarmult_vartime2() {
+/// r = [s1]p1 + [s2]p2
+#[no_mangle]
+pub unsafe extern "C" fn ge25519_double_scalarmult_vartime2(
+    r: *mut Ge25519,
+    p1: *const Ge25519,
+    s1: *const Bignum256Modm,
+    p2: *const Ge25519,
+    s2: *const Bignum256Modm,
+) {
+    let p1 = match EdwardsPoint::try_from(&*p1) {
+        Ok(v) => v,
+        Err(_) => return,
+    };
+    let p2 = match EdwardsPoint::try_from(&*p2) {
+        Ok(v) => v,
+        Err(_) => return,
+    };
+
+    let (s1, s2) = (
+        Scalar::from_unpacked_u32(*s1),
+        Scalar::from_unpacked_u32(*s2),
+    );
 
+    let r1 = s1 * p1 + s2 * p2;
+
+    *r = Ge25519::from(&r1);
 }
 
-pub unsafe extern "C" fn ge25519_scalarmult_base_wrapper() {
+/// Convert to compressed curve form
+#[no_mangle]
+pub unsafe extern "C" fn ge25519_pack(r: *mut [u8; 32], p1: *const Ge25519) {
+    let p1 = match EdwardsPoint::try_from(&*p1) {
+        Ok(v) => v,
+        Err(_) => return,
+    };
+
+    let c1 = p1.compress();
 
+    (*r).copy_from_slice(c1.as_bytes());
 }
 
-pub unsafe extern "C" fn ge25519_scalarmult() {
+/// Unpack from compressed curve form
+#[no_mangle]
+pub unsafe extern "C" fn ge25519_unpack_vartime(r: *mut Ge25519, c: *const [u8; 32]) -> c_int {
+    let c1 = CompressedEdwardsY(*c);
+
+    let r1 = match c1.decompress() {
+        Some(v) => v,
+        None => return 0,
+    };
+
+    (*r) = Ge25519::from(&r1);
+
+    return 1;
+}
 
+/// Copy from p to r
+#[no_mangle]
+pub unsafe extern "C" fn ge25519_copy(r: *mut Ge25519, p: *const Ge25519) {
+    (*r).x = (*p).x;
+    (*r).y = (*p).y;
+    (*r).z = (*p).z;
+    (*r).t = (*p).t;
 }
 
-pub unsafe extern "C" fn ge25519_unpack_vartime() {
+// Point from [hash] in variable time
+#[no_mangle]
+pub unsafe extern "C" fn ge25519_fromfe_frombytes_vartime(
```

Comment:
hey @isislovecruft any idea what [`ge25519_fromfe_frombytes_vartime`](https://github.com/trezor/trezor-firmware/blob/f93a8514e8e25b17b02360d9955c0a30999adf68/crypto/ed25519-donna/ed25519-donna-impl-base.c#L640) method is trying to do? it seems very similar to / slightly inverse of [`ge25519_unpack_negative_vartime`](https://github.com/trezor/trezor-firmware/blob/f93a8514e8e25b17b02360d9955c0a30999adf68/crypto/ed25519-donna/ed25519-donna-impl-base.c#L194), but i'm not sure if this maps to a standard operation or if i need to manually implement an equivalent.